### ISSUE TITLE
Let us run Sidekick offline

### ISSUE THREAD

**Issue by yilmazdurmaz** [NativeScript/sidekick-feedback#415]
I started using Nativescript recently and it just became my favorite; Javascript rocks. Then Sidekick came in. What I love about it, beside mains, is its log windows; showing at what stage the build process is, and what my device says while running my app. Especially when my code editor starts a new build, which is set to save file every second.

With cloud build, this becomes much better. But there are times we need to slow down. Let it be running away from distraction, being on travel or simply cut from the internet, we should be able to continue building on local.

I love cloud build system but I also need my local builds too. In its current shape, Sidekick can't start offline whether or not we have a local build system. Sidekick should be on our side as our sidekick while we work on our Nativescript project; not just some delivery boy which happens to live with us but refuses to do some work :)

### Did you verify whether this issue has already been reported here?
Yes.
There are one-liners in issues that Sidekick won't start without a connection.
Plus this is a feature request. 

### Tell us about the problem
Sidekick needs an offline mode.

### Which platform(s) does the issue occur on?
Any

### Provide the following version numbers that the issue occurs with:
Latest as of the date

### Tell us how to recreate the issue in as much detail as possible
1. Disconnect the internet
2. Start the application
3. Watch the error messages flying and Sidekick hanging. 
4. Decide to write a feature request or vote on some if you find any.

### Does this issue happen every time?
Yes

### Send the client logs
You know, you don't need logs here

**Comment by NickIliev** [comment]
@yilmazdurmaz the NativeScript projects are depending on NPM packages (all dependencies/plugins and even the core modules). So in order to be able to run even a local build, you would need to have those packages installed. That is one of the main reasons why offline mode is not viable options - you simply need an internet connection to be able to execute **npm i**

**Comment by yilmazdurmaz** [comment]
@NickIliev unfortunately it seems you have a gap of knowledge about npm cache.

Unless you clean the cache or packages get broken due disk errors, every package installed in projects are cached by npm. One can even add a package by hand to the cache. Then, when used with cached version numbers most, if not all, packages can be 'npm installed' offline. (I even made a package to search installed packages npm-cache-list). Besides there other tools, like old sinopia, to proxy-cache packages with dependencies in first install.
Also, working with fixed versions includes android sdk too. 

If you dont want to expose a forever-offline code, a tokenized mode for 3-5 offline use would do the job very well. For example, I am also a gamer and I can use my steam games in offline mode for some time before it asks to be online again.

**Comment by teobugslayer** [comment]
Hi @yilmazdurmaz,

Thank you for your feedback.

NativeScript Sidekick is an online service with a desktop client. Is has been build to be always online. The team has no plans to implement offline mode because it is not a very often requested feature. 

We see the benefits of the offline mode but it's simply not as easy to add on to the current implementation.

**Comment by aranmohyeddin** [comment]
Can you explain what makes it difficult? Because one can very easily build with tns without internet but sidekick wouldn't even open without internet... before you even try to do anything. Maybe sidekick is built to work online due to some business reasons, can please clarify that?

**Comment by luisjakon** [comment]
Hello, can anyone tell me if any of your code get exposed or uploaded anywhere when running a local build via Sidekick? Thank you!

**Comment by bundyo** [comment]
No, {N} CLI and Sidekick are not uploading anything when a local build is run.
